gapic-generator-java/src/main/java/com/google/api/generator/gapic/composer/resourcename/ResourceNameHelperClassComposer.java

```diff
@@ -160,8 +160,26 @@ private static List<AnnotationNode> createClassAnnotations() {
             .build());
   }
 
-  private static List<TypeNode> createImplementsTypes() {
-    return Arrays.asList(FIXED_TYPESTORE.get("ResourceName"));
+  /**
+   * Builds a list of types to be implemented. If the type has the same name then we use its full
+   * name
+   *
+   * @param implementingClassName class that is implementing the resulting list. Collisions are
+   *     checked
+   * @return
+   */
+  private static List<TypeNode> createImplementsTypes(String implementingClassName) {
+    List<TypeNode> preprocessed = Arrays.asList(FIXED_TYPESTORE.get("ResourceName"));
+    List<TypeNode> result = new ArrayList<>();
+    for (TypeNode typeNode : preprocessed) {
+      if (!typeNode.isPrimitiveType()
+          && typeNode.reference().name().compareTo(implementingClassName) == 0) {
```

Comment:
I think `typeNode.reference().name().equals(implementingClassName)` does the same thing?

Reply:
I switched to this comparison. It is better.